xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java

```diff
@@ -0,0 +1,270 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package io.grpc.xds;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Any;
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.config.endpoint.v3.ClusterLoadAssignment;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.envoyproxy.envoy.config.route.v3.RouteConfiguration;
+import io.envoyproxy.envoy.service.discovery.v3.AggregatedDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryRequest;
+import io.envoyproxy.envoy.service.discovery.v3.DiscoveryResponse;
+import io.grpc.SynchronizationContext;
+import io.grpc.stub.StreamObserver;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+class XdsTestControlPlaneService extends
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceImplBase {
+  private static final Logger logger = Logger.getLogger(XdsE2eTest.class.getName());
+
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          logger.log(Level.SEVERE, "Exception!" + e);
+        }
+      });
+
+  private static final String ADS_TYPE_URL_LDS =
+      "type.googleapis.com/envoy.config.listener.v3.Listener";
+  private static final String ADS_TYPE_URL_RDS =
+      "type.googleapis.com/envoy.config.route.v3.RouteConfiguration";
+  private static final String ADS_TYPE_URL_CDS =
+      "type.googleapis.com/envoy.config.cluster.v3.Cluster";
+  private static final String ADS_TYPE_URL_EDS =
+      "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
+  private ImmutableMap<String, Listener> ldsResources = ImmutableMap.of();
+  private ImmutableMap<String, RouteConfiguration> rdsResources = ImmutableMap.of();
+  private ImmutableMap<String, Cluster> cdsResources = ImmutableMap.of();
+  private ImmutableMap<String, ClusterLoadAssignment> edsResources = ImmutableMap.of();
+
+  private int ldsVersion = 1;
+  private int rdsVersion = 1;
+  private int cdsVersion = 1;
+  private int edsVersion = 1;
+  private int ldsNonce = 0;
+  private int rdsNonce = 0;
+  private int cdsNonce = 0;
+  private int edsNonce = 0;
+
+  /**
+   * Set static xds configurations, normally set config must be done before sending xds requests.
+   */
+  public void setConfig(XdsTestControlPlaneConfig config) {
```

Comment:
I was not expecting a static, fixed, all-at-once API like this. I was expecting a mutation API where resources can be changed (added, replaced, deleted), and for the simple cases you can just add all the resources (one-by-one) before starting the xds client.

This implementation has a lot of boilerplate, but it really doesn't help us much toward such a purpose. I expect we would want to abstract the concept of a resource type and not duplicate the code for each resource type. We will need some hacks to handle state-of-the-world vs not, but that only applies to one piece of the entire thing.

Reply:
Actually... the hack for state-of-the-world vs not can just be to treat all the resource types as state-of-the-world (send back all resources of a particular type when any of them change). That prevents us from doing some fine-grained testing in xdsClient, but that doesn't seem to be a problem (we can do that more manually for those few tests). It also means we don't have to track which clients are interested in which specific resources.